In [1]:
# https://huggingface.co/datasets/allganize/rag-ko

from huggingface_hub import snapshot_download

snapshot_download(
    repo_id='allganize/rag-ko',
    repo_type='dataset',
    local_dir='./res/rag-ko',
    local_dir_use_symlinks=False
)

print()

/Users/jaehyungpark/Documents/holiday-mission_박재형/mission4_박재형/.venv/lib/python3.13/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
/Users/jaehyungpark/Documents/holiday-mission_박재형/mission4_박재형/.venv/lib/python3.13/site-packages/huggingface_hub/file_download.py:982: UserWarning: `local_dir_use_symlinks` parameter is deprecated and will be ignored. The process to download files to a local folder has been updated and do not rely on symlinks anymore. You only need to pass a destination folder as`local_dir`.
For more details, check out https://huggingface.co/docs/huggingface_hub/main/en/guides/download#download-files-to-local-folder.
  warnings.warn(
Fetching 4 files: 100%|██████████| 4/4 [00:00<00:00, 3187.16it/s]

## 📦 데이터셋 다운로드

Hugging Face에서 `allganize/rag-ko` 데이터셋을 다운로드합니다.

- **`snapshot_download`**: 허깅페이스에서 데이터셋 전체를 로컬에 저장
- **`repo_id`**: 다운로드할 데이터셋 이름
- **`local_dir`**: 저장할 폴더 경로 (`./res/rag-ko`)
- **결과**: RAG(검색 증강 생성)용 한국어 Q&A 데이터셋이 로컬에 저장됨

In [2]:
import pandas as pd


df = pd.read_parquet('./res/rag-ko/data/test-00000-of-00001.parquet')
df

index                                             system  \
0        0  You are a financial expert.\nYou are fluent at...   
1        1  You are a financial expert.\nYou are fluent at...   
2        2  You are a financial expert.\nYou are fluent at...   
3        3  You are a financial expert.\nYou are fluent at...   
4        4  You are a financial expert.\nYou are fluent at...   
..     ...                                                ...   
195    195  You are a financial expert.\nYou are fluent at...   
196    196  You are a financial expert.\nYou are fluent at...   
197    197  You are a financial expert.\nYou are fluent at...   
198    198  You are a financial expert.\nYou are fluent at...   
199    199  You are a financial expert.\nYou are fluent at...   

                                                 human  \
0                           글로벌 저금리 현상이 부각된 원인은 무엇인가요?   
1                       고수익-고위험 부문으로의 쏠림현상에 대해 설명해주세요.   
2    20세기 이후 디지털화의 진전이 금융투자업의 사업모형과 산업구조에 어떤 영향을 미쳤...   
3                    2000년대 이후 글로벌 IB들이 어떤 전략을 추진하였나요?   
4                                금융시스템 개혁법은 언제 제정되었나요?   
..                                                 ...   
195    우리나라 은행들이 기업대출을 통해 외부자금을 필요로 하는 산업에 어떻게 기여하였나요?   
196        은행의 기업대출이 중소기업의 부가가치 성장에 어떻게 기여하였는지 설명해주세요.   
197                              기업 저축의 역설에 대해 설명해주세요.   
198             명목 경제 성장률이 높은 것이 경기호황을 의미하는 이유는 무엇인가요?   
199        대기업들의 중소기업에 대한 부당행위를 근절하면 어떤 긍정적인 효과가 있을까요?   

                                                answer  answer_position  \
0    글로벌 저금리 현상이 부각된 원인은 여러 가지입니다. 첫째, 2008년 글로벌 금융...                3   
1    고수익-고위험 부문으로의 쏠림현상은 투자자들이 더 높은 수익률을 추구하기 위해 위험...                2   
2    20세기 이후 디지털화의 진전은 금융투자업의 사업모형과 산업구조에 큰 영향을 미쳤습...                1   
3    2000년대 이후 글로벌 IB들은 스스로 전산화를 추진하였으며, 이와 밀접하게 연관...                3   
4                           금융시스템 개혁법은 1998년에 제정되었습니다.                3   
..                                                 ...              ...   
195  우리나라 은행들은 기업대출을 통해 외부자금을 필요로 하는 산업에 자금을 효율적으로 ...                3   
196  은행의 기업대출이 중소기업의 부가가치 성장에 기여하는 방식은 크게 두 가지로 볼 수...                2   
197  기업 저축의 역설이란, 기업들이 지나치게 많은 자금을 저축하면서 경제의 역동성이 상...                3   
198  명목 경제 성장률이 높다는 것은 경기가 호황인 상태를 의미합니다. 이는 경제가 활발...                1   
199  대기업들의 중소기업에 대한 부당행위를 근절하면, 중소기업이 받아야 할 몫을 중소기업...                1   

                    answer_context_title  \
0    132579145651725164_KIFRT2021-02 (2)   
1    132579145651725164_KIFRT2021-02 (2)   
2    132579145651725164_KIFRT2021-02 (2)   
3    132579145651725164_KIFRT2021-02 (2)   
4    132579145651725164_KIFRT2021-02 (2)   
..                                   ...   
195        132859249008797508_FARPT21-06   
196        132859249008797508_FARPT21-06   
197      130328682579485000_KIFFS2013-08   
198      130328682579485000_KIFFS2013-08   
199      130328682579485000_KIFFS2013-08   

                                answer_context_summary  
0    2. 뉴노멀의 도래(신 3저의 도래)\n최근 코로나 사태를 포함하여 2008년 글로...  
1    축소되어 자산운용의 어려움이 가중\n∙ 투자자 입장에서는 원하는 수익률을 얻기 어려...  
2    1. 디지털화의 진전과 금융투자업의 변화 : 20세기 이후\n가. 1950~60년대...  
3    ∙ 또한, 투자은행의 명성에 의존하는 대표적인 업무인 IPO, M&A\n등도 전산화...  
4    및 일본 내 해외투자 수요를 포용하는데 적극적\n일련의 변화과정에서 일본 금융당국의...  
..                                                 ...  
195  ... 긍정적인 측면과 부정적인 측면\n∙ 우리나라 은행의 기업대출이 사업체 수 증...  
196  ... 데\n은행의 기업대출 대부분도 중소기업대출이어서 과연 은행이 중소기업대출을\...  
197  ... 현재의 기업 저축률은 불필요하게 높은 수준일 가능성을 시사\n4. 2004년...  
198  ...  비해 명목 성장률이 특별히 높았던 기간이 아니었다. 50) 먼저 2001년...  
199  ... \n여전함을 적나라하게 보여 주고 있다. 62) 이러한 관행은 또한 중소기업...  

[200 rows x 7 columns]

## 📊 데이터 불러오기

다운로드한 Parquet 파일을 pandas DataFrame으로 읽어옵니다.

- **`pd.read_parquet()`**: Parquet 형식의 파일을 DataFrame으로 변환
- **Parquet**: 엑셀처럼 표 형태로 데이터를 저장하는 효율적인 파일 형식
- **`df`**: DataFrame의 약자로, 행과 열로 구성된 표 데이터
- **출력 결과**: 200개의 질문-답변 쌍과 관련 컨텍스트 정보가 표시됨

In [ ]:
print(df.iloc[0]['system'])

## 🔍 첫 번째 데이터 확인

첫 번째 행의 `system` 컬럼 내용을 출력합니다.

- **`df.iloc[0]`**: DataFrame의 첫 번째 행(0번 인덱스) 선택
- **`['system']`**: 해당 행의 `system` 컬럼 값 가져오기
- **`system` 컬럼**: AI에게 주어지는 시스템 프롬프트 (역할 설정 + 컨텍스트 정보)

In [3]:
import re

def extract_contexts(system_prompt):
    context_match = re.search(r'CONTEXT="""\n(.*?)"""', system_prompt, re.DOTALL)
    if not context_match:
        return []

    context_text = context_match.group(1)

    contexts = re.findall(r'\(context \d+\)=(.*?)(?=\n\(context \d+\)=|\Z)', context_text, re.DOTALL)

    cleaned_contexts = []
    for context in contexts:
        lines = context.split('\n')
        cleaned_lines = [line for line in lines if not line.strip().startswith('Title:')]
        cleaned_context = '\n'.join(cleaned_lines).strip()
        cleaned_contexts.append(cleaned_context)

    return cleaned_contexts

system_prompt = df.iloc[0]['system']

extracted_contexts = extract_contexts(system_prompt)

for i, context in enumerate(extracted_contexts, 1):
    print(f"Context {i}:")
    print(context)
    print("-" * 50)

Context 1:
... 변화의 동인은 크게 두 가지로 분류할
수 있다. 하나는 금융업계 전반에 부각되는 디지털화(digitalization)1)이고
다른 하나는 저금리·저성장으로 대표되는 뉴노멀(new normal)이다.  ... . 뉴노멀의 특징으로는 금융투자
및 자산운용의 어려움이 가중되는 경향, 고수익-고위험 부문으로의 쏠림
현상 등의 위험요인과 함께 자산관리서비스 수요의 증가 등 대응 노력을
살펴본다. 제III장은 디지털화 및 뉴노멀에 대응하는 해외 사례들을 살펴본다.  ... 뉴노멀의
주요 특징인 저성장, 저금리 등은 일본의 금융투자업계가 30년 전부터 고민해
온 문제이기 때문이다.  ... 제III장에서 살펴본 해외
사례들에서 살펴본 금융투자업계 변화의 규칙성이 어느 정도 유지될 것이라는
1 디지털화(digitalization)에 대한 ...
--------------------------------------------------
Context 2:
3. 문제점
1) 기업신용위험 평가항목의 적정성
글로벌 금융위기 이후 저금리 기조가 지속되었고, 최근 추가적으로
금리가 하락함에 따라 이자보상배율 등의 지표들이 기업의 부실을
측정하는 데 적정한 지표인지에 대한 검토가 필요
기업신용위험 정 ... 작용할 것으로 예상
저금리 기조가 지속되는 가운데 금융비용이 감소하여 이자보상배
율이 기업의 부실을 가늠하는 지표로 적합한지에 대해 의문이
발생
더불어 코로나19 사태로 인한 중소기업에 대한 금융권 대출 만기
연장 및 이자상환 유예조치로 이 ... 는 추세
이러한 졸업비율 감소 현상의 원인으로 저조한 경기의 영향이
있었던 것으로 추정되지만 개시연도를 기준으로 졸 ...
--------------------------------------------------
Context 3:
2. 뉴노멀의 도래(신 3저의 도래)
최근 코로나 사태를 포함하여 2008년 글로벌 금융위기 이후 전세계적인
저금리·저성장 상황
∙ 글로벌 금융위기를 극복하는 과정에서 미국, 유럽,

## 🔧 컨텍스트 추출 함수

`system` 프롬프트에서 여러 개의 컨텍스트 문단을 추출하는 함수입니다.

### 함수 동작 원리
1. **정규표현식으로 전체 컨텍스트 블록 찾기**: `CONTEXT="""..."""` 부분 추출
2. **개별 컨텍스트 분리**: `(context 1)=...`, `(context 2)=...` 형태로 나눔
3. **Title 줄 제거**: 각 컨텍스트에서 `Title:` 로 시작하는 줄 삭제
4. **정리된 컨텍스트 리스트 반환**

### 실행 결과
- 첫 번째 데이터의 `system` 프롬프트에서 3개의 컨텍스트를 추출
- 각 컨텍스트는 금융 관련 문서 내용

In [ ]:
extracted_contexts[2] == df.iloc[0]['answer_context_summary']

## ✅ 데이터 검증

추출한 컨텍스트가 원본 데이터의 정답 컨텍스트와 일치하는지 확인합니다.

- **`extracted_contexts[2]`**: 추출한 컨텍스트 중 3번째 (인덱스 2)
- **`df.iloc[0]['answer_context_summary']`**: 원본 데이터의 정답 컨텍스트
- **비교 결과**: `True`면 올바르게 추출됨을 의미

In [4]:
contexts = []
for i in range(len(df)):
    system_prompt = df.iloc[i]['system']
    extracted_contexts = extract_contexts(system_prompt)
    contexts.append(extracted_contexts)

## 🔄 전체 데이터 컨텍스트 추출

200개 모든 데이터에서 컨텍스트를 추출하여 리스트에 저장합니다.

- **`for i in range(len(df))`**: 모든 행을 순회 (0~199)
- **`extract_contexts()`**: 각 행의 `system` 프롬프트에서 컨텍스트 추출
- **`contexts.append()`**: 추출한 컨텍스트를 리스트에 추가
- **결과**: 200개의 질문마다 여러 개의 컨텍스트가 담긴 2차원 리스트

In [5]:
questions = [df.iloc[i]['human'] for i in range(len(df))]
contexts_answers_idxs = [df.iloc[i]['answer_position'] - 1 for i in range(len(df))]
contexts_answers = [df.iloc[i]['answer_context_summary'] for i in range(len(df))]
answers = [df.iloc[i]['answer'] for i in range(len(df))]

## 📝 필요한 데이터 추출

RAG 시스템에 필요한 4가지 핵심 정보를 추출합니다.

- **`questions`**: 사용자가 물어보는 질문들 (`human` 컬럼)
- **`contexts_answers_idxs`**: 정답이 들어있는 컨텍스트의 위치 (1부터 시작 → 0부터 시작으로 변환)
- **`contexts_answers`**: 정답이 들어있는 컨텍스트 문단 (`answer_context_summary` 컬럼)
- **`answers`**: 질문에 대한 정답 (`answer` 컬럼)

각각 200개의 요소를 가진 리스트로 저장됩니다.

In [6]:
import pickle

rag_data = {
    'questions': questions,
    'contexts': contexts,
    'contexts_answer_idx': contexts_answers_idxs,
    'contexts_answers': contexts_answers,
    'answers': answers
}

with open('./res/rag_data.pkl', 'wb') as f:
    pickle.dump(rag_data, f)

## 💾 데이터 저장 (Pickle)

추출한 모든 데이터를 딕셔너리로 묶어서 파일로 저장합니다.

- **`rag_data` 딕셔너리**: 5가지 핵심 정보를 key-value 형태로 저장
  - `questions`: 질문 리스트
  - `contexts`: 모든 컨텍스트 리스트 (2차원)
  - `contexts_answer_idx`: 정답 컨텍스트 위치
  - `contexts_answers`: 정답 컨텍스트 내용
  - `answers`: 정답 리스트

- **`pickle.dump()`**: 파이썬 객체를 파일로 저장 (나중에 빠르게 불러오기 가능)
- **저장 경로**: `./res/rag_data.pkl`